### ISSUE TITLE
Suggestion: Add rustanalyzer integration

### ISSUE THREAD

**Issue by CGMossa** [google/evcxr#102]
This would solidify the REPL-experience on Jupyter notebooks.

I don't know how this could be achieved, or what the necessary steps. If someone knows, then
we could initiate a conversation here.

If this is not appropriate, feel free to close this.

**Comment by davidlattimore** [comment]
There's #3 for adding tab completion. Out of curiosity, what would be the other top feature requests? It's possible that each feature of rust analyzer might require separate work to integrate.

I have previously looked at trying to use rust analyzer to do tab completion. I had trouble. I think perhaps I needed to do something else to get it to load the standard library because the completions it was returning didn't seem to include any standard library stuff.

I'm also not sure if the approach that I was trying to take was the right one. I was trying to use rust analyzer directly as a rust library, but perhaps I should have been starting the rust analyzer language server, then connecting to that to make requests. I do like the idea of just using it as a library though, since then it could just be built into the evcxr binary and wouldn't need a separate installation step. OTOH, if it used the language server, it would make it easier for people to upgrade it separately without me doing new evcxr releases.

**Comment by bjorn3** [comment]
Rust analyzer supports signature help, type info for variables, showing (syntax) errors while typing, goto definition/usage, several assists (inlining/extracting variables, adding imports for not yet imported things, adding derives), semantic syntax highlighting (showing which variables are mutable, rainbow highlighting) and more.

**Comment by CGMossa** [comment]
@davidlattimore I don't know what would be the most sensible. I'd vote for the server-approach, mainly because I assume that there would be features that would lean into this kind of setup, as rustanalyzer is used like that with vscode, emacs, etc. But there is an advantage to the baked-in features. Maybe both?

I'm very happy that this wasn't outright dismissed, thank you!

**Comment by davidlattimore** [comment]
It looks like language server protocol (LSP) support for Jupyter Lab has come quite a long way since last I looked. See [krassowski/jupyterlab-lsp](https://github.com/krassowski/jupyterlab-lsp).

It wouldn't help the REPL, but it would be awesome to have something like that working with the Jupyter Kernel.

I don't think it would work to have jupyterlab-lsp connect directly to a Rust language server, since it wouldn't handle :dep commands, existing variables, writing statements outside of functions etc. So Evcxr would need to act as an LSP proxy, altering the requests as they pass through and the responses on the way back.

**Comment by jgarvin** [comment]
I wonder if emacs with ein (it's a great emacs mode for viewing jupyter notebooks, even supports inline images) would be easier to get working. rustic-mode already talks to lsp/analyzer and from ein's perspective each cell can be running the mode for that cell's language.

**Comment by matklad** [comment]
>I think perhaps I needed to do something else to get it to load the standard library 

Yes, you need to explictely test rust-analyzer where the *sources* of standard library are (as we are not processing rlibs at the moment). There's no docs for it, but here's code that configures stdlib from `rustc ---print sysroot` for non-interractive use-case: https://github.com/rust-analyzer/rust-analyzer/blob/e00b9ca3c94d8f88cc5aede562365025a59af06b/crates/rust-analyzer/src/cli/load_cargo.rs

>I was trying to use rust analyzer directly as a rust library, but perhaps I should have been starting the rust analyzer language server, then connecting to that to make requests.

In theory, using it as a library is better, as it is specifically designed as a library with a somewhat well-defined public API (`ra_ide` crate). In practice however, the API is not stable yet, and we are not ready to provide stability guarantees. So, "library API" is a lodestar at the moment, and not a reality. 

In contrast, the LSP protocol specifies stable inter process API, and using it will protect you from our internal refactorings.

**Comment by joellidin** [comment]
Hi, I found this project recently and it looks really cool. I tried it out but I am missing some of the LSP features when using the notebooks. Mostly LSP signatures and documentation for the different functions and methods. Is there a way to get these features or is this a WIP? When searching the internet I just keep coming back to this issue.

**Comment by bionicles** [comment]
hey, i'd use this right now, really enjoying your project here and switched to rust notebooks but i miss rust-analyzer and wish it were easy to run that in this context on the .ipynb files

**Comment by Davidos533** [comment]
Need it!
